src/hotspot/cpu/x86/macroAssembler_x86.cpp

```diff
@@ -9461,5 +9439,55 @@ void MacroAssembler::get_thread(Register thread) {
   }
 }
 
-
 #endif // !WIN32 || _LP64
+
+void MacroAssembler::fast_lock_impl(Register obj, Register hdr, Register thread, Register tmp1, Register tmp2, Label& slow) {
+  assert(hdr == rax, "header must be in rax for cmpxchg");
+  assert_different_registers(obj, hdr, thread, tmp1, tmp2);
+
+  // First we need to check if the lock-stack has room for pushing the object reference.
+  movptr(tmp1, Address(thread, Thread::lock_stack_current_offset()));
+  cmpptr(tmp1, Address(thread, Thread::lock_stack_limit_offset()));
+  jcc(Assembler::greaterEqual, slow);
+
+  Register locked_hdr = tmp2->is_valid() ? tmp2 : tmp1;
+  // Now we attempt to take the fast-lock.
+  // hdr lowest two bits are either 00 or 01 here. Make it 01.
+  orptr(hdr, markWord::unlocked_value);
+  movptr(locked_hdr, hdr);
+  // Clear lowest two bits: we have 01 (see above), now flip the lowest to get 00.
+  xorptr(locked_hdr, markWord::unlocked_value);
```

Comment:
So you want to compute 1) the header but with low order bits are zero, and 2) the header but low order bits are 01.
I think I would compute the first one by using bitwise and with -4. Then I would compute the second by taking the first value + 1. The benefit is that you don't have to rely on the low order bits being either 00 or 01 in the header then, making it more future proof. Somebody might want to use 10 for something, for example.

Reply:
10 is already used for monitors. Flipping the lowest bit would make it 11 - a value that should be impossible in this scenario, and which would make the CAS fail. However, what you suggest is clearer and doesn't rely on implicit knowledge, I'll do that, instead. Thank you!